### ISSUE TITLE
Use NonEmptyChunk for the inputs of ZTransducers and ZSinks

### ISSUE THREAD

**Issue by simpadjo** [zio/zio#3672]
NonEmptyChunk is a wrapper around Chunk with extra allocation and indirection. 
It could be a newtype instead (using tags or at least AnyVal).
Should be pretty much straightforward. Are there any known downsides? 

If it is done, we can switch Conduits to operate on NonEmptyChunks and get squashing of empty chunks for free.

@iravid @jdegoes

**Comment by iravid** [comment]
Not sure using NonEmptyChunk throughout is possible. `ZStream#drain` relies on being able to empty the chunks and is a fairly important operator for control flow.

**Comment by simpadjo** [comment]
I see no problems with it so far. `ZStream.drain` is morally equivalent to `stream.flatMap(_ => ZStream.empty)` and the later will just work.

`mapChunks` wiil accept `NonEmptyChunk` and return `Chunk` at a cost of slightly more complicated implementation.

**Comment by iravid** [comment]
Yes that's possible, yet unfortunately much less efficient :-) Let's see how much we actually gain by doing this. I am not convinced yet.

**Comment by simpadjo** [comment]
No, I don't mean changing `drain` to `flatMap`.
`drain` implementation will still be using `mapChunks`. 
No performance overhead.

**Comment by iravid** [comment]
Ok, let's see where that gets us!

**Comment by simpadjo** [comment]
@iravid here is the prototype: https://github.com/zio/zio/pull/3676
I changed input types of Transducer and Sink to `NonEmptyChunk`. Output type of transducer stays the same.

I wanted to change Stream's output type to be `NonEmptyChunk` but Barbara Liskov said that it should stay. Anyway, if all inputs are `NonEmptyChunk`, we have a type-level proof of not sending empty packets over the conduit.

**Comment by iravid** [comment]
@simpadjo Looks promising! Can you get the branch to a state where it's without the *New/*Old stuff?

**Comment by simpadjo** [comment]
@iravid There are few combinators that I was too lazy to rewrite right now. So I use `*New` and `*Old` functions to convert between old and new representations and piggy-back existing code.

I'd like to do the actually port them when we agree on the plan and have zero-cost NonEmptyChunk.
But there are couple of things I can cleanup, sure.

**Comment by iravid** [comment]
Ah that's fair. I think I'm 👍 on this with the transducer/sink inputs being non-empty. I can't see any immediate problem with that. However with changes like these, it's usually when you get to port everything that you spot the actual problems.

So 👍 from me on pushing forwards with this and let's see where we end up.

It'll definitely be easier writing transducers and sinks with this change!

**Comment by hearnadam** [comment]
I'm not sure this is worth pursuing or backwards compatible.
